org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/debug/FqnConverter.java

```diff
@@ -0,0 +1,311 @@
+        source = classFile.getSource();
+      } else {
+        ICompilationUnit unit = outerClass.getCompilationUnit();
+        source = unit.getSource();
+      }
+
+      Document document = new Document(source);
+      IRegion region = document.getLineInformation(lineNumber);
+      int start = region.getOffset();
+      int end = start + region.getLength();
+
+      iMember = binSearch(outerClass, start, end);
+    } catch (JavaModelException e) {
+      throw new IllegalArgumentException(
+          format(
+              "Unable to find source for class with fqn '%s' in the project '%s'",
+              fileUri, javaProject),
+          e);
+    } catch (BadLocationException e) {
+      throw new IllegalArgumentException("Unable to calculate breakpoint location", e);
+    }
+
+    if (iMember instanceof IType) {
+      return ((IType) iMember).getFullyQualifiedName();
+    }
+
+    if (iMember != null) {
+      fqn = ((IMember) iMember).getDeclaringType().getFullyQualifiedName();
+      while (iMember != null && !(iMember instanceof CompilationUnit)) {
+        if (iMember instanceof SourceType
+            && !((SourceType) iMember).isAnonymous()
+            && iMember.getParent() instanceof SourceMethod) {
+
+          fqn = fqn.replace("$" + iMember.getElementName(), "$1" + iMember.getElementName());
+        }
+
+        iMember = iMember.getParent();
+      }
+
+      return fqn;
+    }
+
+    throw new IllegalArgumentException("Unable to calculate breakpoint location");
+  }
+
+  /** Converts fqn into {@link LocationParameters}. */
+  public static List<LocationParameters> fqnToLocation(
+      List<Object> parameters, IProgressMonitor pm) {
+    Preconditions.checkArgument(parameters.size() >= 2, "Fqn and line number are expected.");
+
+    String fqn = (String) parameters.get(0);
+    Integer lineNumber = Integer.parseInt(parameters.get(1).toString());
+
+    if (pm.isCanceled()) {
+      throw new OperationCanceledException();
+    }
+
+    Pair<char[][], char[][]> fqnPair = prepareFqnToSearch(fqn);
+    List<IType> types;
+    try {
+      types =
+          findTypeByFqn(fqnPair.getKey(), fqnPair.getValue(), SearchEngine.createWorkspaceScope());
+    } catch (JavaModelException e) {
+      throw new RuntimeException(e);
+    }
+
+    if (types.isEmpty()) {
+      throw new RuntimeException("Type with fully qualified name: " + fqn + " was not found");
+    }
+
+    IType type = types.get(0); // TODO we need handle few result! It's temporary solution.
+    if (type.isBinary()) {
+      IClassFile classFile = type.getClassFile();
+      int libId = classFile.getAncestor(IPackageFragmentRoot.PACKAGE_FRAGMENT_ROOT).hashCode();
+      return Collections.singletonList(new LocationParameters(fqn, libId, lineNumber));
+    } else {
+      IPath unitPath = type.getCompilationUnit().getPath();
+      IPath projectPath = type.getJavaProject().getPath();
+
+      URI resourseUri =
```

Comment:
You can use:
https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java#L512